File: android/app/src/main/kotlin/com/vyooo/insta360/pipeline/FrameStage.kt
package com.vyooo.insta360.pipeline

/**
 * A single deterministic transformation in the capture-side pipeline.
 *
 * AI may guide parameters, but a stage always performs the transform itself. A stage returns the
 * (possibly mutated) frame, or `null` to drop it (e.g. the temporal-dedup stage enforcing a 1-in-N
 * frame budget).
 */
interface FrameStage {
    val name: String

    /** Transform [frame] using optional [hints]; return it, or `null` to drop the frame. */
    fun process(frame: PipelineFrame, hints: PipelineHints): PipelineFrame?
}

File: ios/Runner/Insta360/Pipeline/FrameStage.swift
import Foundation

/// A single deterministic transformation in the capture-side pipeline.
/// Swift port of Android `FrameStage.kt`.
///
/// AI may guide parameters, but a stage always performs the transform itself. A stage returns the
/// (possibly mutated) frame, or `nil` to drop it (e.g. the temporal-dedup stage enforcing a frame
/// budget).
protocol FrameStage: AnyObject {
    var name: String { get }

    /// Transform `frame` using optional `hints`; return it, or `nil` to drop the frame.
    /// Declared `throws` so `FramePipeline` can fall open (pass-through) on any stage error, mirroring
    /// the Android try/catch. Concrete stages that never throw simply omit `throws`.
    func process(_ frame: PipelineFrame, hints: PipelineHints) throws -> PipelineFrame?
}
